proposals/XXXX-extend-property-wrappers-to-function-and-closure-parameters.md

```diff
@@ -0,0 +1,388 @@
+          // when the user taps the item. 
+          // Unfortunately, to mutate the item
+          // we have to manually index our
+          // data source.
+          shoppingItems[index].quanity += 1
+        }
+      
+    }
+  }
+}
+```
+
+With an appropriate initializer we would be able to simplify the above code, therefore reducing boilerplate:
+
+```swift
+struct MyView: View {
+  @State 
+  private var shoppingItems: [Item]
+
+  var body: some View {
+    ForEach($shoppingItems) { @Binding shoppingItem in
+    
+      Text(shoppingItem.name)
+        .onTapGesture {
+          shoppingItem.quanity += 1
+        }
+      
+    }
+  }
+}
+```
+
+
+## Proposed solution
+
+We propose to extend the contexts were application of property-wrapper types is allowed. Namely, application of such types will be allowed on function and closure parameters:
+
+```swift
+@propertyWrapper
+struct Clamped<Value: Comparable> {
+  ...
+    
+  var projectedValue: Self {
+    self
+  }
+}
+
+func increment(
+  @Clamped(to: 0 ... 100) percent: Int = 100
+) { ... }
+
+myPercentage
+  .modify(inSeconds: 3) { @Clamped percent in
+    percent = 100 
+  }
+```
+
+
+## Detailed design
+
+Property wrappers are essentially sugar wrapping a given property with compiler synthesized code. This proposal retains this principle employing the following rules for transformation.
+
+### Property Wrappers on Function Parameters
+
+Function parameters marked with a set of property wrapper custom attributes must conform to the following rules:
+
+1. Each property wrapper type must have a suitable `init(wrappedValue:)` for initializing the property wrapper from an instance of its `wrappedValue` type.
+2. Each `wrappedValue` getter must be `nonmutating`.
+3. Default values must be expressed in terms of the innermost `wrappedValue` type.
+
+The transformation of a property wrapper parameter will take place is as follows:
+
+1. The external parameter name will remain unchanged.
+2. The internal parameter name will be prefixed with an underscore, and the type of this parameter is the backing property wrapper type.
+3. A local computed property representing  `wrappedValue` will be synthesized by the compiler and named per the original (non-prefixed) parameter name. The accessors will mirror the `wrappedValue` accessors. A setter will only be synthesized for the local property if the `wrappedValue` setter is `nonmutating`, or if the wrapper is a reference type.
+4. If the property wrapper defines a `projectedValue`, a local computed property representing  `projectedValue` will be synthesized by the compiler and named per the original parameter name prefixed with a dollar sign (`$`). The same accessor rules for `wrappedValue` apply to `projectedValue`.
+5. When passing an argument to a property wrapper parameter, the compiler will wrap the argument in the appropriate `init(wrappedValue:)` call.
+
+#### Transformation Example:
+
+```swift
+@propertyWrapper
+struct Percentage {
```

Comment:
I'm not attached to this example - I wanted something relatively simple that didn't need additional arguments in the wrapper attribute (which ruled out `Clamped`). @filip-sakel please feel free to change this if you think of something better 🙂

Reply:
How would you feel about a `@Prefixed` wrapper:

```swift
@propertyWrapper
struct Prefixed {

  var prefixValue: String
  var originalValue: String

 
  init(
    wrappedValue originalValue: String,
    by prefixValue: String
  ) { ... }

   
  var wrappedValue: String { 
    get { ... }
  }

  var projectedValue: Self { ... }

}
```

So that we can write:

```swift
func log(
  @Prefixed(by: "Error: ") error: String
) { ... }


log(error: "Couldn't fetch images.")
// prints: Error: Couldn't fetch images.

log(_error: Prefixed(by: "UIError: ", wrappedValue: "Couldn't fetch images."))
// prints: UIError: Couldn't fetch images.
```

I'm not sure if that's any better. If `@Prefixed` isn't good enough we could explore something error-related. Anyway, let me know what you think of `@Prefixed`.

Reply:
`@Prefixed` still requires additional arguments in the wrapper attribute (the `(by: "Error")` part), and I'm leaning toward not allowing arguments in the wrapper attribute initially with this proposal

Reply:
Right, I just started writing the wrapper and completely forgot the single-argument part 😅. I'll give this some more thought to see if I can come up with something better than `@Percentage`.

Reply:
A simple example might be a property wrapper for when you always want to be given a lowercased string.

```swift
func example(@Lowercased _ urlScheme: String) -> Void { ... }
```

Would the `Lowercased` initializer be allowed to have a `with locale: Locale? = nil` default argument?

Reply:
@benrimmington these are great questions. Could you please ask them on the pitch thread?

Reply:
@hborla what do you think of `@Lowercased`?

In my opinion, we can be more creative with `@Lowercased` as the current example is supposed to report some progress, whereas we could use lowercase for something like a username or the 'urlScheme' example that @benrimmington proposed.

Reply:
I like that example!